core/arch/arm/mm/mobj.c

```diff
@@ -0,0 +1,432 @@
+	moph->mobj.ops = &mobj_phys_ops;
+	moph->pa = pa;
+	moph->va = (vaddr_t)va;
+
+	return &moph->mobj;
+}
+
+/*
+ * mobj_virt implementation
+ */
+
+static void mobj_virt_assert_type(struct mobj *mobj);
+
+static void *mobj_virt_get_va(struct mobj *mobj, size_t offset)
+{
+	mobj_virt_assert_type(mobj);
+
+	return (void *)(vaddr_t)offset;
+}
+
+static void mobj_virt_free(struct mobj *mobj)
+{
+	mobj_virt_assert_type(mobj);
+	free(mobj);
+}
+
+static const struct mobj_ops mobj_virt_ops __rodata_unpaged = {
+	.get_va = mobj_virt_get_va,
+	.free = mobj_virt_free,
+};
+
+static void mobj_virt_assert_type(struct mobj *mobj __maybe_unused)
+{
+	assert(mobj->ops == &mobj_virt_ops);
+}
+
+struct mobj mobj_virt = { .ops = &mobj_phys_ops, .size = SIZE_MAX };
+
+/*
+ * mobj_mm implementation
+ */
+
+struct mobj_mm {
+	tee_mm_entry_t *mm;
+	struct mobj *parent_mobj;
+	struct mobj mobj;
+};
+
+static struct mobj_mm *to_mobj_mm(struct mobj *mobj);
+
+static size_t mobj_mm_offs(struct mobj *mobj, size_t offs)
+{
+	tee_mm_entry_t *mm = to_mobj_mm(mobj)->mm;
+
+	return (mm->offset << mm->pool->shift) + offs;
+}
+
+static void *mobj_mm_get_va(struct mobj *mobj, size_t offs)
+{
+	return mobj_get_va(to_mobj_mm(mobj)->parent_mobj,
+			   mobj_mm_offs(mobj, offs));
+}
+
+
+static TEE_Result mobj_mm_get_pa(struct mobj *mobj, size_t offs,
+				    size_t granule, paddr_t *pa)
+{
+	return mobj_get_pa(to_mobj_mm(mobj)->parent_mobj,
+			   mobj_mm_offs(mobj, offs), granule, pa);
+}
+
+static TEE_Result mobj_mm_get_cattr(struct mobj *mobj, uint32_t *cattr)
+{
+	return mobj_get_cattr(to_mobj_mm(mobj)->parent_mobj, cattr);
+}
+
+static bool mobj_mm_matches(struct mobj *mobj, enum buf_is_attr attr)
+{
+	return mobj_matches(to_mobj_mm(mobj)->parent_mobj, attr);
+}
+
+static void mobj_mm_free(struct mobj *mobj)
+{
+	struct mobj_mm *m = to_mobj_mm(mobj);
+
+	tee_mm_free(m->mm);
+	free(m);
+}
+
+static const struct mobj_ops mobj_mm_ops __rodata_unpaged = {
+	.get_va = mobj_mm_get_va,
+	.get_pa = mobj_mm_get_pa,
+	.get_cattr = mobj_mm_get_cattr,
+	.matches = mobj_mm_matches,
+	.free = mobj_mm_free,
+};
+
+static struct mobj_mm *to_mobj_mm(struct mobj *mobj)
+{
+	assert(mobj->ops == &mobj_mm_ops);
+	return container_of(mobj, struct mobj_mm, mobj);
+}
+
+struct mobj *mobj_mm_alloc(struct mobj *mobj_parent, size_t size,
+			      tee_mm_pool_t *pool)
+{
+	struct mobj_mm *m = calloc(1, sizeof(*m));
+
+	if (!m)
+		return NULL;
+
+	m->mm = tee_mm_alloc(pool, size);
+	if (!m->mm) {
+		free(m);
+		return NULL;
+	}
+
+	m->parent_mobj = mobj_parent;
+	m->mobj.size = size;
+	m->mobj.ops = &mobj_mm_ops;
+
+	return &m->mobj;
+}
+
+#ifdef CFG_PAGED_USER_TA
+/*
+ * mobj_paged implementation
+ */
+
+static void mobj_paged_free(struct mobj *mobj);
+
+static const struct mobj_ops mobj_paged_ops __rodata_unpaged = {
+	.free = mobj_paged_free,
+};
+
+static void mobj_paged_free(struct mobj *mobj)
+{
+	assert(mobj->ops == &mobj_paged_ops);
+	free(mobj);
+}
+
+struct mobj *mobj_paged_alloc(size_t size)
```

Comment:
In `.c` file `mobj_paged_alloc` is inside `#ifdef CFG_PAGED_USER_TA`. But in `.h` file it is outside that `#ifdef`